### PR TITLE
Add e2e test for upgrade

### PR DESCRIPTION
**Pull request by yusuoh**
1.Add a new e2e test for upgrade. It tries to fetch a base version and upgrade it to the latest version.
2.Add some util functions for download release

### CODE REVIEW

File: tests/e2e/framework/framework_test.go [hunk 1]
```diff
@@ -53,7 +53,7 @@ func newCommonConfig(testID string) (*CommonConfig, error) {
 	if err != nil {
 		return nil, err
 	}
-	k, err := newKubeInfo(t.TempDir, t.RunID)
+	k, err := newKubeInfo(t.TempDir, t.RunID, "")
```
Comment (chxchx) [inline tests/e2e/framework/framework_test.go]: Can we have another function called `newKubeInfoWithBaseVersion(tmpDir, runID, baseVersion string)` and keep the original `newKubeInfo(tmpDir, runID string)` which calls `newKubeInfoWithBaseVersion()` with the empty string as baseVersion?

File: tests/e2e/framework/kubernetes.go [hunk 2]
```diff
@@ -80,10 +80,16 @@ type KubeInfo struct {
...
+	// Release directory
+	ReleaseDir string
+	// Use baseversion if not empty.
+	BaseVersion string
 }
 
 // newKubeInfo create a new KubeInfo by given temp dir and runID
-func newKubeInfo(tmpDir, runID string) (*KubeInfo, error) {
+// If baseVersion is not empty, will use the specified release of Istio instead of the local one.
+func newKubeInfo(tmpDir, runID string, baseVersion string) (*KubeInfo, error) {
```
Comment (chxchx) [inline tests/e2e/framework/kubernetes.go]: runID string -> runID // string is redundant

File: tests/e2e/tests/upgrade/upgrade_test.go [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+		return nil, err
+	}
+	for _, c := range cookies {
+		// Set cookie
+		req.AddCookie(&c)
+	}
+	return client.Do(req)
+}
+
+func closeResponseBody(r *http.Response) {
```
Comment (chxchx) [inline tests/e2e/tests/upgrade/upgrade_test.go]: I see many duplicated functions from bookinfo test, such as inspect, closeRespoinseBody, deleteRules, applyRules, etc. Perhaps it is better to put them in a shared utility.

File: tests/util/commonUtils.go [hunk 4]
```diff
@@ -246,3 +265,73 @@ func GetResourcePath(p string) string {
...
+
+// ExtractTarGz extracts a .tar.gz file into current dir.
+func ExtractTarGz(gzipStream io.Reader) error {
+	uncompressedStream, err := gzip.NewReader(gzipStream)
+	if err != nil {
+		return errors.Wrap(err, "Fail to uncompress")
+	}
+	tarReader := tar.NewReader(uncompressedStream)
+
+	for true {
```
Comment (chxchx) [inline tests/util/commonUtils.go]: `true` can be removed

File: tests/e2e/tests/upgrade/upgrade_test.go (hunk 3 above)
Comment (yusuoh) [reply tests/e2e/tests/upgrade/upgrade_test.go]: Those are specific for bookinfo app, so maybe not suitable for general test utils.

File: tests/istio.mk [hunk 5]
```diff
@@ -66,4 +66,7 @@ e2e_mixer: istioctl generate_yaml
 e2e_bookinfo: istioctl generate_yaml
 	go test -v -timeout 20m ./tests/e2e/tests/bookinfo -args ${E2E_ARGS} ${EXTRA_E2E_ARGS}
 
+e2e_upgrade: istioctl generate_yaml
+	go test -v -timeout 20m ./tests/e2e/tests/upgrade -args ${E2E_ARGS} ${EXTRA_E2E_ARGS}
+
 e2e_all: e2e_simple e2e_mixer e2e_bookinfo
```
[the remark this answers is not in the document]
Comment (yusuoh) [reply tests/istio.mk]: I plan to add it later once the test case is more complete. Currently it is better to be run manually.

File: tests/e2e/framework/framework_test.go (hunk 1 above)
Comment (yusuoh) [reply tests/e2e/framework/framework_test.go]: Done

**Comment by yusuoh** [comment]
Thanks for the review. PTAL

File: tests/e2e/framework/framework_test.go (hunk 1 above)
Comment (yusuoh) [reply tests/e2e/framework/framework_test.go]: Reverted as changing that caused a lint error (newKubeInfo with empty string is not used by any code)

**Comment by andraxylia** [comment]
cc @myidpt

**Pull request merged**